osu.Game/Rulesets/UI/Scrolling/Algorithms/IScrollAlgorithm.cs

```diff
@@ -6,15 +6,33 @@ namespace osu.Game.Rulesets.UI.Scrolling.Algorithms
     public interface IScrollAlgorithm
     {
         /// <summary>
-        /// Given a point in time, computes the time at which it enters the time range.
+        /// Given a point in time associated with an object's origin
+        /// and the spatial distance between the edge and the origin of the object along the scrolling axis,
+        /// computes the time at which the object initially enters the time range.
         /// </summary>
-        /// <remarks>
-        /// E.g. For a constant time range of 5000ms, the time at which t=7000ms enters the time range is 2000ms.
-        /// </remarks>
-        /// <param name="time">The point in time.</param>
+        /// <example>
+        /// Let's assume the following parameters:
+        /// <list type="bullet">
+        ///     <item><paramref name="originTime"/> = 7000ms,</item>
+        ///     <item><paramref name="offset"/> = 100px,</item>
+        ///     <item><paramref name="timeRange"/> = 5000ms,</item>
+        ///     <item><paramref name="scrollLength"/> = 1000px</item>
+        /// </list>
+        /// and a constant scrolling rate.
+        /// To arrive at the end of the scrolling container, the object's origin has to cover
+        /// <code>1000 + 100 = 1100px</code>
+        /// so that the edge starts at the end of the scrolling container.
+        /// One scroll length of 1000px covers 5000ms of time, so the time required to cover 1100px is equal to
+        /// <code>5000 * (1100 / 1000) = 5500ms,</code>
+        /// and therefore the object should start being visible at
+        /// <code>7000 - 5500 = 1500ms.</code>
```

Comment:
Hmm, this example is a little bit hard to understand, although the correct result is still arrived at. Specifically the line `the origin has to cover 1100px` to me assumes that the edge of the hitobject is already positioned at the edge of the scroll area, but this is not true as they are separated by 300 more pixels (`= (7000 - 5000) * pixels-per-ms - offset`).

```
Consider the following:
                          
< < < < < < < < < < < < < < < < < < < < < < < < (Scroll direction)
         Scroll Area              Object
[--------------------------]      [-----o-----]
      TimeRange = 5000                Origin
       Distance = 1000            Offset = 100px
                                    Time = 7000

Pixels-per-ms (ppms)     = 1000 / 5000 = 0.2
Distance-to-origin (dto) = ppms * 7000 = 1400
Distance-to-edge (dte)   = dto - 100 = 1300

Visible-at = dte * 1 / ppms - TimeRange = 1500
```

Am I missing something? Because this also means that the implementation of `OverlappingScrollAlgorithm` should instead be:

`(originTime - TimeRange - offset * timeRange / scrollLength) / controlPoint.Multiplier)`

That is to say, take the origin time delta, subtract the offset duration, and finally factor in the time-range manipulation control point.

Reply:
As far as I can tell, both lines of reasoning are correct. The point of possible confusion is, that when I wrote the 1100px number, what I meant is that the hit object will travel that much in *screen space*, not absolutely from `time = 0`. This is why my assumption of the edge being at the start of the scroll wasn't an incorrect one to make.

The only thing that's not quite right in your formula is that you've divided too much by the control point multiplier (it should not affect the origin time). In fact, if you rearrange your formula with that fixed, you actually get the same result:
```
originTime - TimeRange / controlPoint.Multiplier - offset * timeRange / scrollLength / controlPoint.Multiplier =
(extract common factor)
originTime - TimeRange / controlPoint.Multiplier * (1 + offset / scroll length) =
(move division by scrollLength outside inner bracket)
originTime - TimeRange / controlPoint.Multiplier * (scrollLength + offset) / scrollLength
```
at which point differs from my formula only with the order of operations (I do multiplications first in code so that divisions don't introduce floating point inaccuracies when dividing early).

Please let me know how I can make this clearer.

Reply:
Ok thanks! I had a feeling it was just my seeing the issue in a different way :)